<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Adam Koziol" /><link rel="canonical" href="https://OLC-Bioinformatics.github.io/AlleleFinder/stec_combined_subunits/" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Combined STEC subunits - OLC Bioinformatics AlleleFinder</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Combined STEC subunits";
        var mkdocs_page_input_path = "stec_combined_subunits.md";
        var mkdocs_page_url = "/AlleleFinder/stec_combined_subunits/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> OLC Bioinformatics AlleleFinder
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../installation/">Installation</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../downloads/">Downloads</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../profile_reduce/">Profile Reduction</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../allele_translate_reduce/">Allele Translation/Reduction</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../allele_find/">Allele Finding</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../allele_translate_find/">Allele Translate/Finding</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../aa_allele_find/">Amino Acid Allele Finding</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../allele_split/">Allele Splitting</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../allele_concatenate/">Allele Concatenating</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="#">Combined STEC subunits</a>
    <ul class="current">
    </ul>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">OLC Bioinformatics AlleleFinder</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Combined STEC subunits</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/OLC-Bioinformatics/AlleleFinder/edit/main/docs/stec_combined_subunits.md">Edit on AlleleFinder</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h2 id="table-of-contents">Table of Contents</h2>
<ol>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#prerequisites">Prerequisites</a></li>
<li><a href="#inputs">Inputs</a></li>
<li><a href="#running-the-script">Running the Script</a></li>
<li><a href="#command-line-arguments">Command Line Arguments</a></li>
<li><a href="#outputs">Outputs</a></li>
<li><a href="#interpreting-the-results">Interpreting the Results</a></li>
<li><a href="#troubleshooting">Troubleshooting</a></li>
<li><a href="#additional-resources">Additional Resources</a></li>
</ol>
<h2 id="introduction">Introduction <a name="introduction"></a></h2>
<p>This script processes STEC alleles with both A and B subunits. It performs BLAST analyses to identify the best matching alleles for each subunit, detects novel alleles, and generates reports summarizing the results. This is essential for accurate subunit-level typing and characterization of Shiga toxin-producing <em>Escherichia coli</em> (STEC).</p>
<h2 id="prerequisites">Prerequisites <a name="prerequisites"></a></h2>
<p>Before running the script, ensure you have the following:</p>
<ol>
<li>Query files in FASTA format containing nucleotide or amino acid sequences.</li>
<li>Allele database files for the relevant subunits (A and B) in FASTA format.</li>
</ol>
<h2 id="inputs">Inputs <a name="inputs"></a></h2>
<p>The script requires the following inputs:</p>
<ol>
<li>Query files in FASTA format.</li>
<li>Allele database files for the A and B subunits (e.g., <code>Stx1A_aa_*.fasta</code>, <code>Stx1B_aa_*.fasta</code>, <code>Stx2A_aa_*.fasta</code>, <code>Stx2B_aa_*.fasta</code>).</li>
<li>Optionally, a directory containing split AA allele databases for subunit-level BLAST analyses.</li>
</ol>
<h2 id="running-the-script">Running the Script <a name="running-the-script"></a></h2>
<p>To run the script, use the following command:</p>
<pre><code class="language-bash">stec.py stec_combined_subunits --allele_path /path/to/allele_folder --report_path /path/to/output_folder --query_path /path/to/query_folder --blast_mode blastn --split_aa_db_dir /path/to/split_aa_db_dir
</code></pre>
<h2 id="command-line-arguments">Command Line Arguments <a name="command-line-arguments"></a></h2>
<p>The script accepts the following command line arguments:</p>
<ul>
<li><code>-h, --help</code>: Show this help message and exit.</li>
<li><code>--version</code>: Show program's version number and exit.</li>
<li><code>-a, --allele_path</code>: Specify the path to the folder containing allele files. Default: <code>alleles</code> in the current working directory.</li>
<li><code>-r, --report_path</code>: Specify the path to the folder for output reports. Default: <code>reports</code> in the current working directory.</li>
<li><code>-q, --query_path</code>: Specify the path to the folder containing query files in FASTA format. Default: <code>query</code> in the current working directory.</li>
<li><code>--blast_mode</code>: Choose BLAST mode: <code>blastn</code>, <code>tblastx</code>, <code>blastx</code>, or <code>blastn+tblastx</code>. Default: <code>blastx</code>.</li>
<li><code>--preliminary</code>: Run a preliminary screen using blastn (do not run additional downstream analyses).</li>
<li><code>--split_aa_db_dir</code>: Path to directory containing split AA allele databases (required for full analysis).</li>
<li><code>-n, --num_alignments</code>: Number of alignments to return. Default: 10.</li>
<li><code>-c, --cutoff</code>: Percent identity cutoff. Default: 99.0.</li>
<li><code>-v, --verbosity</code>: Set the logging level. Options: debug, info, warning, error, critical. Default: info.</li>
</ul>
<h2 id="outputs">Outputs <a name="outputs"></a></h2>
<p>The script generates the following output files:</p>
<ul>
<li><code>novel_alleles.fasta</code>: FASTA file containing all novel alleles discovered.</li>
<li><code>stec_combined_nt_report.tsv</code>: TSV report summarizing nucleotide-level matches and novel alleles.</li>
<li><code>stec_combined_aa_report.tsv</code>: TSV report summarizing amino acid-level matches and novel alleles.</li>
<li><code>stec_combined_nt_aa_report.tsv</code>: Combined report for both nucleotide and amino acid analyses.</li>
<li>Individual FASTA files for novel subunits and operons, as appropriate.</li>
</ul>
<h2 id="interpreting-the-results">Interpreting the Results <a name="interpreting-the-results"></a></h2>
<p>The main reports (<code>stec_combined_nt_report.tsv</code>, <code>stec_combined_aa_report.tsv</code>, and <code>stec_combined_nt_aa_report.tsv</code>) summarize the best matches for each query, percent identity, and any novel alleles detected. Additional notes may indicate partial matches, ambiguous bases, or gap events.</p>
<h2 id="troubleshooting">Troubleshooting <a name="troubleshooting"></a></h2>
<p>If you encounter issues:</p>
<ul>
<li>Ensure all input files are in the correct format and location.</li>
<li>Confirm that BLAST+ is installed and accessible from your command line.</li>
<li>Check that the split AA database directory contains the required subunit FASTA files.</li>
<li>Review error messages for missing files or permission issues.</li>
</ul>
<h2 id="additional-resources">Additional Resources <a name="additional-resources"></a></h2>
<p>For more information, see the <a href="https://olc-bioinformatics.github.io/AlleleFinder/stec_combined_subunits"><code>stec_combined_subunits</code></a> documentation or open an issue on <a href="https://github.com/OLC-Bioinformatics/AlleleFinder/issues">GitHub</a>.</p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../allele_concatenate/" class="btn btn-neutral float-left" title="Allele Concatenating"><span class="icon icon-circle-arrow-left"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>Copyright © 2025, Government of Canada</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/OLC-Bioinformatics/AlleleFinder/" class="fa fa-code-fork" style="color: #fcfcfc"> AlleleFinder</a>
        </span>
    
    
      <span><a href="../allele_concatenate/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
